src/packages/emmett-mongodb/src/eventStore/mongoDBEventStore.ts

```diff
@@ -154,89 +210,159 @@ export class MongoDBEventStore implements EventStore<number> {
-      const currentStream = await this.collection.findOne({
-        streamName: { $eq: streamName },
-      });
+      const currentStream = await this.collection.findOne(
+        { streamName: { $eq: streamName } },
+        { useBigInt64: true },
+      );
       throw new ExpectedVersionConflictError(
-        currentStream?.events.length ?? -1,
-        stream.events.length,
+        currentStream?.metadata?.streamPosition ?? 0n,
+        stream.metadata.streamPosition,
       );
     }
 
-    const { streamType, streamId } = fromStreamName(streamName);
-
-    await executeProjections(
-      {
-        streamName,
-        streamType,
-        streamId,
-        streamVersion: updatedStream.events.length,
-        events: updatedStream.events,
-      },
-      options?.projections,
-    );
-
     return {
-      nextExpectedStreamVersion: updatedStream.events.length,
+      nextExpectedStreamVersion: updatedStream.metadata.streamPosition,
       createdNewStream,
     };
   }
 }
 
-export const getMongoDBEventStore = (collection: Collection<EventStream>) => {
-  const eventStore = new MongoDBEventStore(collection);
+export const getMongoDBEventStore = (
+  options: ConstructorParameters<typeof MongoDBEventStore>[0],
+) => {
+  const eventStore = new MongoDBEventStore(options);
   return eventStore;
 };
 
-function executeProjections<EventType extends Event>(
-  params: StreamToProject<EventType>,
-  projections?: Array<Projection<EventType>>,
-) {
-  return Promise.all((projections ?? []).map((project) => project(params)));
+async function handleAsyncProjections<
+  EventType extends Event = Event,
+>(options: {
+  streamName: StreamName;
+  events: ReadEvent<EventType, ReadEventMetadata>[];
+  projections: MongoDBAsyncProjection[];
+  collection: Collection<EventStream>;
+}) {
+  const projections = filterProjections(options.projections, options.events);
+  for (const { handle } of projections) {
+    await handle(options.events, {
+      streamName: options.streamName,
+      collection: options.collection,
+    });
+  }
+}
+
+function filterProjections<
+  Projection extends MongoDBAsyncProjection | MongoDBInlineProjection,
+>(projections: Projection[], events: ReadEvent[]) {
+  const eventTypes = events.map((e) => e.type);
+  const filteredProjections = projections
+    .map((p) => p.projection)
+    .filter((p) => p.canHandle.some((t) => eventTypes.includes(t)));
+  return filteredProjections;
+}
+
+export function mongoDBInlineProjection<
+  EventType extends Event,
+  ShortInfoType extends DefaultRecord,
+>(options: {
+  name: string;
+  canHandle: CanHandle<EventType>;
+  evolve: (state: ShortInfoType, event: EventType) => ShortInfoType;
+}): MongoDBInlineProjection {
+  return {
+    type: 'inline',
+    projection: {
+      name: options.name,
+      canHandle: options.canHandle,
+      handle: async (events, { streamName, collection }) => {
+        const stream = await collection.findOne({ streamName });
```

Comment:
We should not be loading anything additionally in inline projection, as we already should have already this data in the stream state.

Reply:
Think I might need your help on this for inline projections. I can't seem to figure out how we can have the projection handler be passed the stream state if we run the inline projections before the update so that we can have a single update instead of multiple when appending to the stream.

It feels like there's kind of a dependency circle between the handler needing to know the initial state, either current short projection on doc (requires a read), or the inline projections would be passed all events on the stream and not just the ones about to append, but that would need to be done after the stream update meaining a second update required for the projection handler if that makes sense.

Reply:
I'll adjust the type definition in https://github.com/event-driven-io/emmett/pull/132. I think that the best way will be to return inline projections together with the event stream and allow the reading result to be passed to append as it was (e.g. ` { expectedStreamState: ReadResult }` then we can read that from read result and assume that's the current projection state. Together with optimistic concurrency and atomic update, it should be sufficient.

For that, I'll need to adjust the command handler and event store definition to allow the extended event stream result to be taken (specific for event store type).